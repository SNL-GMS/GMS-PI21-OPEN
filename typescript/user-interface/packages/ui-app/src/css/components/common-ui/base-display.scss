.base-display {
  display: flex;
  position: relative;
  flex-direction: column;
  overflow: hidden;
  width: 100%;
  contain: strict; // Performance optimization to speed up render step
  &.full-width-height {
    height: 100%;
    width: 100%;
    flex-grow: 1;
  }
  &.scroll-box--x {
    overflow-x: auto;
    overflow-y: hidden;
  }
  &.scroll-box--y {
    overflow-x: hidden;
    overflow-y: auto;
  }
}
