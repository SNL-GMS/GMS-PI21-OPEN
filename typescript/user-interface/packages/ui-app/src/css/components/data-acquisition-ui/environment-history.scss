.environment-history-display {
  &__chart-container {
    height: auto;
    width: 100%;
    overflow-x: auto;
  }
  .external-axis {
    min-height: 50px;
    background-color: var(--gms-prominent-background, $gms-prominent-background);
    border-radius: $rounded-rectangle-radius;
    overflow-y: hidden;
    overflow-x: auto;
  }
}
