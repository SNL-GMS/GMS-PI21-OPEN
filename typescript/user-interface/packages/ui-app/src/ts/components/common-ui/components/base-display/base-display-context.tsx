import type GoldenLayout from '@gms/golden-layout';
import React from 'react';

/**
 * Used for providing access to the base display size and a reference to the
 * golden layout container.
 */
export interface BaseDisplayContextData {
  /**
   * A reference to the golden layout container
   */
  glContainer: GoldenLayout.Container;

  /**
   * the width of the display in px, inclusive of the padding on the sides
   */
  widthPx: number;

  /**
   * the height of the display in px, inclusive of the padding on the sides
   */
  heightPx: number;
}

/**
 * Instantiate the Base Display Context and set up the defaults as undefined.
 */
export const BaseDisplayContext: React.Context<BaseDisplayContextData> = React.createContext<
  BaseDisplayContextData
>(undefined);
