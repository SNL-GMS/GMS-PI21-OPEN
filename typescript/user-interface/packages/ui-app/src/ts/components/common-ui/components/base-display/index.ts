export * from './base-display';
export * from './base-display-context';
