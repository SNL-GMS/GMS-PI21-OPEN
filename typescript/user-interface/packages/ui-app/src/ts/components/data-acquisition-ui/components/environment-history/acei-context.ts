import type { SohTypes } from '@gms/common-model';
import React from 'react';

export type AceiMonitorTypeOption = 'CHOOSE_A_MONITOR_TYPE' | SohTypes.AceiType;

/** ACEI selected context data */
export interface AceiContextData {
  selectedAceiType: SohTypes.AceiType;
  setSelectedAceiType(aceiType: SohTypes.AceiType): void;
}

/** ACEI context, provides the selected ACEI type and the function to select the ACEI type */
export const AceiContext: React.Context<AceiContextData> = React.createContext<AceiContextData>(
  undefined
);
