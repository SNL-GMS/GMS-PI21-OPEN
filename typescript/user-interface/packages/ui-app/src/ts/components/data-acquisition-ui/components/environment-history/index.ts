export { EnvironmentContainer as SohEnvironmentHistory } from './environment-history-container';
