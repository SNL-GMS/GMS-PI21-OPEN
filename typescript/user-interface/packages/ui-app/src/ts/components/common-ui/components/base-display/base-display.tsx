/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/prop-types */
import { getDataAttributesFromProps, useElementSize, useForceUpdate } from '@gms/ui-util';
import defer from 'lodash/defer';
import * as React from 'react';

import { TabContextMenu } from '~analyst-ui/common/menus/tab-context-menu';

import type { BaseDisplayContextData } from './base-display-context';
import { BaseDisplayContext } from './base-display-context';
import type { BaseDisplayProps } from './types';

const useForceUpdateOnFirstRender = () => {
  const forceUpdate = useForceUpdate();
  React.useEffect(() => {
    defer(forceUpdate);
    // !FIX ESLINT Validate and check REACT HOOK dependencies
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
};

/**
 * A base display that should be at the base of all display components.
 * Adds consistent padding to each display, and exposes the width and height
 * of the display in the BaseDisplayContext.
 *
 * @param props requires a reference to the glContainer.
 * Also accepts data attributes in the form 'data-cy': 'example-component'
 */
export const BaseDisplay: React.FunctionComponent<React.PropsWithChildren<
  BaseDisplayProps
  // eslint-disable-next-line react/function-component-definition
>> = props => {
  /**
   * Base display size behavior
   */
  const [displayRef, heightPx, widthPx] = useElementSize();

  /**
   * On the very first mount, call forceUpdate so that height and width will
   * propagate to the consumers of the context.
   */
  useForceUpdateOnFirstRender();

  /**
   * the context menu handler, if provided
   */
  // eslint-disable-next-line @typescript-eslint/unbound-method
  const { onContextMenu, tabName } = props;

  /**
   * Get any data attributes provided to this display (like data-cy attributes)
   */
  const dataAttributes = getDataAttributesFromProps(props);

  const baseDisplayContextData: BaseDisplayContextData = React.useMemo(
    () => ({
      glContainer: props.glContainer,
      widthPx: props.glContainer?.width ?? widthPx ?? 0,
      heightPx: props.glContainer?.height ?? heightPx ?? 0
    }),
    [heightPx, props.glContainer, widthPx]
  );

  return (
    <div
      className={`base-display ${props.className ?? ''}`}
      ref={ref => {
        displayRef.current = ref;
      }}
      onContextMenu={onContextMenu}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...dataAttributes}
    >
      <BaseDisplayContext.Provider value={baseDisplayContextData}>
        {tabName ? <TabContextMenu tabName={tabName} /> : null}
        {props.children}
      </BaseDisplayContext.Provider>
    </div>
  );
};
