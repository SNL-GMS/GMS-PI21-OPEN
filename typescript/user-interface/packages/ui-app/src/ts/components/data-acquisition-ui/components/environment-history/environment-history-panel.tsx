/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/prop-types */
import { SohTypes } from '@gms/common-model';
import { prettifyAllCapsEnumType } from '@gms/common-util';
import { DeprecatedToolbarTypes } from '@gms/ui-core-components';
import { useGetHistoricalAceiDataQuery } from '@gms/ui-state';
import * as React from 'react';

import { useBaseDisplaySize } from '~components/common-ui/components/base-display/base-display-hooks';
import { messageConfig } from '~components/data-acquisition-ui/config/message-config';
import { DrillDownTitle } from '~components/data-acquisition-ui/shared/drill-down-components';
import { BaseToolbar } from '~components/data-acquisition-ui/shared/toolbars/base-toolbar';
import { useTrendTimeIntervalSelector } from '~components/data-acquisition-ui/shared/toolbars/trend-time-interval-hook';
import {
  convertSohMonitorTypeToAceiMonitorType,
  isAnalogAceiMonitorType
} from '~components/data-acquisition-ui/shared/utils';

import type { AceiMonitorTypeOption } from './acei-context';
import { AceiContext } from './acei-context';
import { validateNonIdealState } from './non-ideal-states';
import { WeavessDisplay } from './weavess-display';

export interface EnvironmentHistoryPanelProps {
  station: SohTypes.UiStationSoh;
  channelSohs: SohTypes.ChannelSoh[];
  sohHistoricalDurations: number[];
}

/**
 * @returns a list of all environmental issues from the SohMonitorType enum
 */
const getEnvSohMonitorTypeNames = () =>
  Object.keys(SohTypes.SohMonitorType).filter(SohTypes.isEnvironmentalIssue);

/**
 * TODO: add support for a  default in the toolbar dropdown
 *
 * @returns undefined if the option matches the default. Otherwise, return the enum value.
 */
const convertAceiMonitorTypeOptionToEnum = (option: AceiMonitorTypeOption): SohTypes.AceiType =>
  option && option !== 'CHOOSE_A_MONITOR_TYPE' ? SohTypes.AceiType[option] : undefined;

/**
 * Creates and manages state for the selector dropdown list for Env Monitor Types.
 * The dropdown  will contain all monitor types from the SohMonitorType enum that are ACEI monitors.
 * Modifies the AceiContext to update which monitor type has been selected.
 *
 * @returns an array containing exactly two items. The first is a dropdown,
 * the second is the currently selected option from the dropdown, or undefined
 * if no ACEI monitor type is selected.
 */
const useMonitorTypeDropdown = (): [DeprecatedToolbarTypes.DropdownItem, SohTypes.AceiType] => {
  const envMonitorTypes: AceiMonitorTypeOption[] = [
    'CHOOSE_A_MONITOR_TYPE', // default
    ...getEnvSohMonitorTypeNames()
      .map(convertSohMonitorTypeToAceiMonitorType)
      // TODO: remove filtering once analog types are supported
      .filter(monitorType => !isAnalogAceiMonitorType(monitorType))
  ];

  const context = React.useContext(AceiContext);

  const dropdown: DeprecatedToolbarTypes.DropdownItem = {
    label: 'Select Monitor Type',
    rank: undefined,
    tooltip: 'Select Monitor Type',
    type: DeprecatedToolbarTypes.ToolbarItemType.Dropdown,
    dropdownOptions: envMonitorTypes,
    dropdownText: envMonitorTypes.map(monitorType => prettifyAllCapsEnumType(monitorType, false)),
    value: context.selectedAceiType ? context.selectedAceiType : undefined,
    onChange: value => context.setSelectedAceiType(value),
    widthPx: 220
  };

  const aceiMonitorType = convertAceiMonitorTypeOptionToEnum(context.selectedAceiType);

  return [dropdown, aceiMonitorType];
};

/**
 * The Environment History Panel
 * Composes together the toolbar and the env history charts.
 * Performs a query when a monitor type is selected (or if it is passed down). Then renders the
 * line charts based on the result.
 * Renders a non-ideal state if the query doesn't get the required data in the result.
 */
// eslint-disable-next-line react/function-component-definition
export const EnvironmentHistoryPanel: React.FunctionComponent<EnvironmentHistoryPanelProps> = props => {
  const [startTimeMS, endTimeMs, timeIntervalSelector] = useTrendTimeIntervalSelector(
    'ACEI',
    props.sohHistoricalDurations
  );

  const [monitorTypeDropdown, selectedMonitorType] = useMonitorTypeDropdown();
  const [widthPx] = useBaseDisplaySize();
  const ONE_HUNDRED = 100;
  const queryInput: SohTypes.UiHistoricalAceiInput = {
    endTime: endTimeMs / 1000,
    startTime: startTimeMS / 1000,
    stationName: props.station.stationName,
    type: selectedMonitorType
  };

  const historicalAceiQuery = useGetHistoricalAceiDataQuery(queryInput);
  // update loading state for the button that shows this panel
  const nonIdealState = validateNonIdealState(
    selectedMonitorType,
    historicalAceiQuery.isLoading,
    startTimeMS,
    endTimeMs,
    historicalAceiQuery.data
  );
  return (
    <>
      <BaseToolbar
        widthPx={widthPx}
        itemsLeft={[monitorTypeDropdown]}
        itemsRight={[timeIntervalSelector]}
      />
      {selectedMonitorType ? (
        <DrillDownTitle
          title={props.station.stationName}
          subtitle={prettifyAllCapsEnumType(selectedMonitorType)}
          description={messageConfig.labels.decimationDescription(ONE_HUNDRED)}
        />
      ) : undefined}

      {nonIdealState || (
        <WeavessDisplay
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...props}
          startTimeMs={startTimeMS}
          endTimeMs={endTimeMs}
          aceiData={historicalAceiQuery.data}
        />
      )}
    </>
  );
};
