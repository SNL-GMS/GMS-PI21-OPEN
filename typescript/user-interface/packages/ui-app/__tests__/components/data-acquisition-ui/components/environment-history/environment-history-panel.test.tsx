/* eslint-disable react/jsx-no-constructed-context-values */
import { SohTypes } from '@gms/common-model';
import { sohConfiguration } from '@gms/ui-state/__tests__/__data__/soh-configuration-query-data';
import { render } from '@testing-library/react';
import React from 'react';

import { BaseDisplayContext } from '../../../../../src/ts/components/common-ui/components/base-display/base-display-context';
import { AceiContext } from '../../../../../src/ts/components/data-acquisition-ui/components/environment-history/acei-context';
import { EnvironmentHistoryPanel } from '../../../../../src/ts/components/data-acquisition-ui/components/environment-history/environment-history-panel';
import { testStationSoh } from '../../../../__data__/data-acquisition-ui/soh-overview-data';

const MOCK_TIME = 1611153271425;
const MOCK_TIME_STR = '2021-01-20 02:34:31';

const mockDate: any = new Date(MOCK_TIME);
mockDate.now = () => MOCK_TIME;
Date.constructor = jest.fn(() => new Date(MOCK_TIME));
jest.spyOn(global, 'Date').mockImplementation(() => mockDate);
Date.now = jest.fn(() => MOCK_TIME);
Date.UTC = jest.fn(() => MOCK_TIME);

jest.mock('moment-precise-range-plugin', () => {
  return {};
});

jest.mock('moment', () => {
  // mock chain builder pattern
  const mMoment = {
    utc: jest.fn(() => mMoment),
    format: jest.fn(() => MOCK_TIME_STR)
  };

  // mock the constructor and to modify instance methods
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const fn: any = jest.fn(() => {
    mMoment.format = jest.fn(() => MOCK_TIME_STR);
    return mMoment;
  });

  // mock moment methods that depend on moment not on a moment instance
  fn.unix = () => ({ utc: () => mMoment });
  return fn;
});

jest.mock('@gms/ui-state', () => {
  const actual = jest.requireActual('@gms/ui-state');
  return {
    ...actual,
    useGetSohConfigurationQuery: jest.fn(() => ({
      data: sohConfiguration,
      isLoading: false
    })),
    useGetHistoricalAceiDataQuery: jest.fn(() => ({
      data: [],
      isLoading: false
    }))
  };
});
describe('Environment history panel', () => {
  it('should be defined', () => {
    expect(Date.now()).toEqual(MOCK_TIME);
    expect(EnvironmentHistoryPanel).toBeDefined();
  });
  it('matches the snapshot empty channel soh data', () => {
    const { container } = render(
      <BaseDisplayContext.Provider
        value={{
          glContainer: undefined,
          heightPx: 100,
          widthPx: 100
        }}
      >
        <AceiContext.Provider
          value={{
            selectedAceiType: SohTypes.AceiType.AMPLIFIER_SATURATION_DETECTED,
            setSelectedAceiType: jest.fn()
          }}
        >
          <EnvironmentHistoryPanel
            channelSohs={[]}
            // eslint-disable-next-line @typescript-eslint/no-magic-numbers
            sohHistoricalDurations={[1000, 5000, 8000]}
            station={testStationSoh}
          />
        </AceiContext.Provider>
      </BaseDisplayContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
