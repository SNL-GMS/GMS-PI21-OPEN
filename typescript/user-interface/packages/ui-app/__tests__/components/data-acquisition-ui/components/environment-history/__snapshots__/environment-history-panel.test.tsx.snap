// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Environment history panel matches the snapshot empty channel soh data 1`] = `
<div>
  <div
    class="toolbar"
    style="width: 60px;"
  >
    <div
      class="toolbar__left-group"
    >
      <div
        class="toolbar-item toolbar-item__left"
      >
        <div>
          <div
            class="dropdown-container"
          >
            <span
              class="dropdown-selector"
              data-cy="select-monitor-type-dropdown"
            >
              <div
                class="bp4-html-select"
              >
                <select
                  style="min-width: 220px; width: 220px;"
                  title="Select Monitor Type"
                >
                  <option
                    value="CHOOSE_A_MONITOR_TYPE"
                  >
                    Choose A Monitor Type
                  </option>
                  <option
                    value="AUTHENTICATION_SEAL_BROKEN"
                  >
                    Authentication Seal Broken
                  </option>
                  <option
                    value="BACKUP_POWER_UNSTABLE"
                  >
                    Backup Power Unstable
                  </option>
                  <option
                    value="CALIBRATION_UNDERWAY"
                  >
                    Calibration Underway
                  </option>
                  <option
                    value="CLIPPED"
                  >
                    Clipped
                  </option>
                  <option
                    value="CLOCK_DIFFERENTIAL_TOO_LARGE"
                  >
                    Clock Differential Too Large
                  </option>
                  <option
                    value="DEAD_SENSOR_CHANNEL"
                  >
                    Dead Sensor Channel
                  </option>
                  <option
                    value="DIGITIZER_ANALOG_INPUT_SHORTED"
                  >
                    Digitizer Analog Input Shorted
                  </option>
                  <option
                    value="DIGITIZER_CALIBRATION_LOOP_BACK"
                  >
                    Digitizer Calibration Loop Back
                  </option>
                  <option
                    value="DIGITIZING_EQUIPMENT_OPEN"
                  >
                    Digitizing Equipment Open
                  </option>
                  <option
                    value="EQUIPMENT_HOUSING_OPEN"
                  >
                    Equipment Housing Open
                  </option>
                  <option
                    value="EQUIPMENT_MOVED"
                  >
                    Equipment Moved
                  </option>
                  <option
                    value="GPS_RECEIVER_OFF"
                  >
                    Gps Receiver Off
                  </option>
                  <option
                    value="GPS_RECEIVER_UNLOCKED"
                  >
                    Gps Receiver Unlocked
                  </option>
                  <option
                    value="MAIN_POWER_FAILURE"
                  >
                    Main Power Failure
                  </option>
                  <option
                    value="VAULT_DOOR_OPENED"
                  >
                    Vault Door Opened
                  </option>
                  <option
                    value="ZEROED_DATA"
                  >
                    Zeroed Data
                  </option>
                </select>
                <span
                  class="bp4-icon bp4-icon-double-caret-vertical"
                  icon="double-caret-vertical"
                >
                  <svg
                    data-icon="double-caret-vertical"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      Open dropdown
                    </desc>
                    <path
                      d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="toolbar__center-group"
    >
      <div
        class="toolbar__whitespace"
        style="width: 44px;"
      />
    </div>
    <div
      class="toolbar__right-group"
    >
      <div
        class="toolbar-item
                    "
      >
        <div
          class="toolbar-button--capsule"
        >
          <div
            class="date-range-picker"
            id="date-range-picker-id-3"
          >
            <label
              class="bp4-label bp4-label bp4-inline date-range-picker__label"
            >
              Time Range: 2021-01-20 02:34:31 to 2021-01-20 02:34:31
            </label>
            <div>
              <button
                class="bp4-button date-range-picker__apply-button"
                data-cy="date-range-picker-edit"
                type="button"
              >
                <span
                  class="bp4-button-text"
                >
                  Edit
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="soh-drill-down-station-label display-title"
    data-cy="drill-down-display-title"
  >
    Test
    <div
      class="display-title__subtitle"
    >
      Amplifier Saturation Detected
       (Displaying 100% of available points)
    </div>
  </div>
  <div
    class="bp4-non-ideal-state"
  >
    <h4
      class="bp4-heading"
    >
      No Data
    </h4>
    <div>
      No historical environmental data
    </div>
  </div>
</div>
`;
